flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mailbox.java

```diff
@@ -20,17 +20,85 @@
 
 import javax.annotation.Nonnull;
 
+import java.util.List;
+
 /**
  * A mailbox is basically a blocking queue for inter-thread message exchange in form of {@link Runnable} objects between
- * multiple producer threads and a single consumer.
+ * multiple producer threads and a single consumer. This has a lifecycle of closed -> open -> (quiesced) -> closed.
  */
 public interface Mailbox extends MailboxReceiver, MailboxSender {
 
 	/**
-	 * The effect of this is that all pending letters are dropped and the given priorityAction
-	 * is enqueued to the head of the mailbox.
+	 * This enum represents the states of the mailbox lifecycle.
+	 */
+	enum State {
+		OPEN, QUIESCED, CLOSED
+	}
+
+	/**
+	 * Open the mailbox. In this state, the mailbox supports put and take operations.
+	 */
+	void open();
+
+	/**
+	 * Quiesce the mailbox. In this state, the mailbox supports only take operations and all pending and future put
+	 * operations will throw {@link MailboxStateException}.
+	 */
+	void quiesce();
+
+	/**
+	 * Close the mailbox. In this state, all pending and future put operations and all pending and future take
+	 * operations will throw {@link MailboxStateException}. Returns all letters that were still enqueued.
+	 *
+	 * @return list with all letters that where enqueued in the mailbox at the time of closing.
+	 */
+	@Nonnull
+	List<Runnable> close();
+
+	/**
+	 * The effect of this is that all pending letters in the mailbox are dropped and the given priorityLetter
+	 * is enqueued to the head of the mailbox. Dropped letters are returned. This method should only be invoked
+	 * by code that has ownership of the mailbox object and only rarely used, e.g. to submit special events like
+	 * shutting down the mailbox loop.
+	 *
+	 * @param priorityLetter action to enqueue atomically after the mailbox was cleared.
+	 * @throws MailboxStateException if the mailbox is quiesced or closed.
+	 */
+	@Nonnull
+	List<Runnable> clearAndPut(@Nonnull Runnable priorityLetter) throws MailboxStateException;
```

Comment:
with `putFirst` to put poisionLetter at the head, do we still need this `clearAndPut`?

Reply:
Main reason is that `putFirst` can block if the queue is currently full and therefore should not be invoked from the task thread, `clearAndPut` never blocks. So there is a distinct use-case for each.